### DISCUSSION TITLE
confused by `ggml_new_tensor_2d` example

### DISCUSSION THREAD

**Discussion by hashibit** [ggml-org/whisper.cpp#1177] [Q&A] [answered]
hi, I'm quite new to the `ggml` library and try to learn it. While reading the document, i came across the following example, and confused by the dimensions.
```
//   {
//       struct ggml_tensor * a = ggml_new_tensor_2d(ctx, GGML_TYPE_F32, 2, 3);
//
//       // a[1, 2] = 1.0f;
//       *(float *) ((char *) a->data + 2*a->nb[1] + 1*a->nb[0]) = 1.0f;
//
//       // a[2, 0] = 2.0f;
//       *(float *) ((char *) a->data + 0*a->nb[1] + 2*a->nb[0]) = 2.0f;
//
//       ...
//   }
```
from my limited understanding, `a` is a two dimension tensor, first dimension is `2`, second dimension is `3`. like `c` array `float a[3][2]`

my question is the following assignment: 

```
//       // a[2, 0] = 2.0f;
//       *(float *) ((char *) a->data + 0*a->nb[1] + 2*a->nb[0]) = 2.0f;
```

 it access the first dimension index with `2`, isn't that an `index-out-of-bounds` operation? like `a[0][2]`.

thanks in advance!

**Comment by ggerganov** [comment, marked as answer]
Oh wow - you right. Sorry about the confusion. Your understanding is the correct one. I'll fix the help

Edit: here is the new version:

https://github.com/ggerganov/ggml/blob/1a5d5f331de1d3c7ace40d86fe2373021a42f9ce/include/ggml/ggml.h#L130-L146
